spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/QueryPushdownIntegrationTestBase.java

```diff
@@ -468,4 +436,549 @@ public void testAggregateExpressions() {
+    assertThat(result.get(0).get(0)).isEqualTo(6);
+  }
+
+  @Test
+  public void testJoinQuery() {
+    writeTestDataToBigQuery(initialData(), testDataset.toString() + "." + testTable);
+    Dataset<Row> df = readTestDataFromBigQuery(testDataset.toString() + "." + testTable);
+    df.createOrReplaceTempView("numStructDF");
+
+    writeTestDataToBigQuery(
+        additionalData(), testDataset.toString() + "." + testTable + "_to_join");
+    Dataset<Row> df_to_join =
+        readTestDataFromBigQuery(testDataset.toString() + "." + testTable + "_to_join");
+    df_to_join.createOrReplaceTempView("numStructDF_to_join");
+
+    List<Row> result =
+        spark
+            .sql(
+                "SELECT "
+                    + "numStructDF.num1 AS a_num1, "
+                    + "numStructDF.num2 AS a_num2, "
+                    + "numStructDF.num3 AS a_num3, "
+                    + "numStructDF_to_join.num1 AS b_num1, "
+                    + "numStructDF_to_join.num2 AS b_num2, "
+                    + "numStructDF_to_join.num3 AS b_num3 "
+                    + "FROM numStructDF RIGHT JOIN numStructDF_to_join "
+                    + "ON numStructDF.num1 = numStructDF_to_join.num2 "
+                    + "WHERE numStructDF_to_join.num2 > 2 ")
+            .collectAsList();
+    /*
+     +------+------+------+------+------+------+
+     |a_num1|a_num2|a_num3|a_num2|b_num2|b_num3|
+     +------+------+------+------+------+------+
+     |  null|  null|  null|     6|     5|     3|
+     |     4|     3|     2|     3|     4|     1|
+     +------+------+------+------+------+------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      if (r.get(0) == null) {
+        assertThat(r.get(4)).isEqualTo(5);
+      } else {
+        assertThat(r.get(0)).isEqualTo(r.get(4));
+      }
+    }
+
+    // disabling pushdown and testing the same
+    BigQueryConnectorUtils.disablePushdownSession(spark);
+    result =
+        spark
+            .sql(
+                "SELECT "
+                    + "numStructDF.num1 AS a_num1, "
+                    + "numStructDF.num2 AS a_num2, "
+                    + "numStructDF.num3 AS a_num3, "
+                    + "numStructDF_to_join.num1 AS b_num1, "
+                    + "numStructDF_to_join.num2 AS b_num2, "
+                    + "numStructDF_to_join.num3 AS b_num3 "
+                    + "FROM numStructDF RIGHT JOIN numStructDF_to_join "
+                    + "ON numStructDF.num1 = numStructDF_to_join.num2 "
+                    + "WHERE numStructDF_to_join.num2 > 2 ")
+            .collectAsList();
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      if (r.get(0) == null) {
+        assertThat(r.get(4)).isEqualTo(5);
+      } else {
+        assertThat(r.get(0)).isEqualTo(r.get(4));
+      }
+    }
+  }
+
+  /** Creating a Dataset of NumStructType which will be used to write to BigQuery */
+  protected Dataset<Row> initialData() {
```

Comment:
Can we improve the method names? Something like `tableData()` and `joinTableData()`?

I'd actually still recommend a static variable. Something like the static final variables we have in TestConstants class